src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexBuilder.cs

```diff
@@ -76,13 +73,19 @@ internal sealed class SymbolicRegexBuilder<TSet> where TSet : IComparable<TSet>,
         // capturing states that have been created
         internal HashSet<DfaMatchingState<TSet>> _capturingStateCache = new();
 
+        internal Dictionary<(SymbolicRegexNode<TSet>, TSet elem, uint context, bool IsBackTracking), SymbolicRegexNode<TSet>> _derivativeCache = new();
```

Comment:
Added a comment with rationale, also pasting here since it's interesting:
```
// The following dictionaries are used as caches for operations that recurse over the structure of SymbolicRegexNode.
// These operations are called potentially on every step of the matching process, and they may do linear work in the
// of the pattern in each call. Thus, caching is necessary to avoid a quadratic worst-case over multiple steps of
// matching when simplification rules fail to eliminate the portions being walked over.
```
Currently with the stress tests we have removing these caches wouldn't be a problem. However, we don't have a proof that our current simplification rules always avoid these worst-cases. There might be hope for removing some of these caches with future work on the theory side.

Reply:
@veanes also tagging you since you might be interested in seeing the comment I added.

Reply:
I fully agree with this comment. But I want to point out that `_derivativeCache` seems absolutely essential. The caches related to _pruning_ are perhaps avoidable (as you say -- with improved understanding on the theory side).